lib/simulator-xcode-11.4.js

```diff
@@ -0,0 +1,189 @@
+   * The simulator could be in both running and shutdown states
+   * in order for this method to run as expected.
+   *
+   * @param {string} payload the content of the PEM certificate
+   * @param {CertificateOptions} opts
+   */
+  async addCertificate (payload, opts = {}) {
+    const {
+      isRoot = true,
+    } = opts;
+    const tmpRoot = await tempDir.openDir();
+    try {
+      const certPath = path.resolve(tmpRoot, 'certificate.pem');
+      log.info(`Installing SSL ${isRoot ? 'root ' : ''}certificate '${_.truncate(payload, {length: 20})}'`);
+      if (isRoot) {
+        await this.simctl.addRootCertificate(certPath);
+      } else {
+        await this.simctl.addCertificate(certPath);
+      }
+    } finally {
+      await fs.rimraf(tmpRoot);
+    }
+  }
+
+  /**
+   * @override
+   * Simulates push notification delivery to the booted simulator
+   *
+   * @since Xcode SDK 11.4
+   * @param {Object} payload - The object that describes Apple push notification content.
+   * It must contain a top-level "Simulator Target Bundle" key with a string value matching
+   * the target application‘s bundle identifier and "aps" key with valid Apple Push Notification values.
+   * For example:
+   * {
+   *   "Simulator Target Bundle": "com.apple.Preferences",
+   *   "aps": {
+   *     "alert": "This is a simulated notification!",
+   *     "badge": 3,
+   *     "sound": "default"
+   *   }
+   * }
+   */
+  async pushNotification (payload) {
+    await this.simctl.pushNotification(payload);
+  }
+
+  /**
+   * @override
+   * Sets the permissions for the particular application bundle.
+   *
+   * @param {string} bundleId - Application bundle identifier.
+   * @param {Object} permissionsMapping - A mapping where kays
+   * are service names and values are their corresponding status values.
+   * The following keys are supported:
+   * - all: Apply the action to all services.
+   * - calendar: Allow access to calendar.
+   * - contacts-limited: Allow access to basic contact info.
+   * - contacts: Allow access to full contact details.
+   * - location: Allow access to location services when app is in use.
+   * - location-always: Allow access to location services at all times.
+   * - photos-add: Allow adding photos to the photo library.
+   * - photos: Allow full access to the photo library.
+   * - media-library: Allow access to the media library.
+   * - microphone: Allow access to audio input.
+   * - motion: Allow access to motion and fitness data.
+   * - reminders: Allow access to reminders.
+   * - siri: Allow use of the app with Siri.
+   * The following values are supported:
+   * - yes: To grant the permission
+   * - no: To revoke the permission
+   * - unset: To reset the permission
+   * @throws {Error} If there was an error while changing permissions.
+   */
+  async setPermissions (bundleId, permissionsMapping) {
+    log.debug(`Setting ${JSON.stringify(permissionsMapping, null, 2)} access for '${bundleId}'`);
```

Comment:
This log entry will be quite difficult to read, with the multiline JSON in the middle
```
Setting {
  "calendar": "1",
  "contacts-limited": "2"
} access for 'io.appium.something'
```

Reply:
do you have any proposals on how to improve the logs?

Reply:
I just wouldn't embed a multi-line thing. The easiest would be 
```
Setting access for '${bundleId}': ${JSON.stringify(permissionsMapping, null, 2)}
```

Reply:
changed